ui/platform_window/x11/x11_window_base.cc

```diff
@@ -534,17 +536,38 @@ void X11WindowBase::OnWMStateUpdated() {
   // Propagate the window minimization information to the client.
   ui::PlatformWindowState state =
       ui::PlatformWindowState::PLATFORM_WINDOW_STATE_NORMAL;
-  if (IsMinimized() != was_minimized) {
-    if (IsMinimized()) {
-      state = ui::PlatformWindowState::PLATFORM_WINDOW_STATE_MINIMIZED;
-    } else if (IsMaximized()) {
-      // When the window is recovered from minimized state, set state to the
-      // previous maximized state if it was like that. Otherwise, NORMAL state
-      // will be set.
-      state = ui::PlatformWindowState::PLATFORM_WINDOW_STATE_MAXIMIZED;
-    }
+  if (IsMaximized())
+    state = ui::PlatformWindowState::PLATFORM_WINDOW_STATE_MAXIMIZED;
+  else if (IsMinimized())
+    state = ui::PlatformWindowState::PLATFORM_WINDOW_STATE_MINIMIZED;
+
+  // |restored_bounds_in_pixels_| can tell if the maximize/restore has been
+  // triggered by client or not. Typically, if it was the client who triggered
+  // that, |restored_bounds_in_pixels_| had values stored. Thus, we can be
+  // sure the client is not flooded with window states when the sources of a
+  // change has been the client itself.
+  if ((restored_bounds_in_pixels_.IsEmpty() &&
+       IsMaximized() != was_maximized) ||
+      IsMinimized() != was_minimized) {
     delegate_->OnWindowStateChanged(state);
+    // A workaround to update the opaque view layout:
+    // In a case when the state of a window has been changed due to some server
+    // side events (window is moved in a such a way so that it becomes
+    // automatically maximized/ restored), the bounds must be sent right after
+    // the state change. But, for example, in gnome with XWayland, the bounds
```

Comment:
question: is this an XWayland behavior, or gnome-shell/x11 also does the same?

Reply:
oh. I rechecked x11 and found a bug. leave this pr for a while. I'll fix it tomorrow and will ask you to review again.

Reply:
both. I remove was_maximized check as well, because in gnome-shell/x11 both restore and maximized are followed after the bounds change, which results in not updated view.